lib/teleterm/daemon/daemon.go

```diff
@@ -63,6 +67,52 @@ func New(cfg Config) (*Service, error) {
 	}, nil
 }
 
+// relogin makes the Electron app display a login modal to trigger re-login.
+func (s *Service) relogin(ctx context.Context, req *api.ReloginRequest) error {
+	// The Electron app cannot display two login modals at the same time, so we have to cut short any
+	// concurrent relogin requests.
+	if !s.reloginMu.TryLock() {
+		return trace.AlreadyExists("another relogin request is in progress")
+	}
+	defer s.reloginMu.Unlock()
+
+	const reloginUserTimeout = time.Minute
+	timeoutCtx, cancelTshdEventsCtx := context.WithTimeout(ctx, reloginUserTimeout)
+	defer cancelTshdEventsCtx()
+
+	if _, err := s.tshdEventsClient.Relogin(timeoutCtx, req); err != nil {
+		if status.Code(err) == codes.DeadlineExceeded {
+			return trace.Wrap(err, "the user did not refresh the session within %s", reloginUserTimeout.String())
+		}
+
+		return trace.Wrap(err, "could not refresh the session")
+	}
+
+	return nil
+}
+
+// retryWithRelogin tries the given function. If the function returns an error that appears to be
+// resolvable with relogin, then it requests relogin and tries the function a second time.
+func (s *Service) retryWithRelogin(ctx context.Context, fn func() error, reloginReq *api.ReloginRequest) error {
```

Comment:
I'd expand the comment to say something like this:

`retryWithRelogin` is reserved for cases where the retryable request does not originate from the Electron app, for example when the request is made a long-running goroutine such as a gateway. When the request originates from the Electron app and daemon.Service is merely an intermediary, the retry flow is handled by `clusters.addMetadataToRetryableError` and the JavaScript version of `client.RetryWithRelogin` with the same name.

---

People who see this function must be able to tell the difference between this function, `client.RetryWithRelogin`, `clusters.addMetadataToRetryableError` and [`retryWithRelogin.ts`](https://github.com/gravitational/teleport/blob/caacc54696529da66185ad7ea8fe45d1053b144c/web/packages/teleterm/src/ui/utils/retryWithRelogin.ts).